PyExpLabSys/drivers/keithley_2400.py

```diff
@@ -128,65 +131,99 @@ def read_voltage(self):
         # Values is: voltage, current, ohm, time, status
         # Only the voltage is measured, current is either
         # NaN or the source-setpoint.
+        print('DEBUG KLAF 2400 Voltage', repr(raw))
         values = raw.split(',')
         voltage = float(values[0])
         # timestamp = float(values[3])
         # print(self._parse_status(values[4]))
         # Also return timestamp?
         return voltage
 
-    def set_source_function(self, function=None):
-        if function in ('i', 'I'):
-            self.scpi_comm('SOURCE:FUNCTION CURRENT')
-        if function in ('v', 'V'):
-            self.scpi_comm('SOURCE:FUNCTION VOLTAGE')
+    def set_source_function(self, function=None, source_range=None):
+        if function.lower() in ('i', 'I'):
+            self.scpi_comm(':SOURCE:FUNCTION CURRENT')
+            self.scpi_comm(':SOURCE:CURRENT:RANGE {}'.format(source_range))
+        if function.lower() in ('v', 'V'):
+            self.scpi_comm(':SOURCE:FUNCTION VOLTAGE')
+            self.scpi_comm(':SOURCE:VOLTAGE:RANGE {}'.format(source_range))
+
         actual_function = self.scpi_comm('SOURCE:FUNCTION?')
         return actual_function
 
+    def set_sense_function(self, function, sense_range=None):
+        """
+        Set the sense range, a value of None returns the current value without
+        changing the actual value. A range value of 0 indicates auto-range.
+        """
+        if function.lower() in ('i' 'current'):
+            self.scpi_comm(':SENSE:FUNCTION:ON "CURRENT"')
+            if sense_range == 0:
+                self.scpi_comm(':SENSE:CURRENT:RANGE:AUTO ON')
+            else:
+                self.scpi_comm(':SENSE:CURRENT:RANGE {}'.format(sense_range))
+        if function.lower() in ('v', 'voltage'):
+            #  TODO: Configure read-back!!!
+            self.scpi_comm(':SENSE:FUNCTION:ON "VOLTAGE"')
+            if sense_range == 0:
+                self.scpi_comm(':SENSE:VOLTAGE:RANGE:AUTO ON')
+            else:
+                self.scpi_comm(':SENSE:VOLTAGE:RANGE {}'.format(sense_range))
+        self.scpi_comm(':SENSE:FUNCTION:ON?')
```

Comment:
Is it on purpose this is sent twice to the instrument, but the first response discarded?